File: solutions/cpp/grains/1/grains.cpp
#include "grains.h"

namespace grains {
    unsigned long long square(int num){
        if (num == 1) return 1;
        unsigned long long sum = 1;
        
        for (int i = 2; i < num+1; ++i)
            sum += sum;
        return sum;
    }
    unsigned long long total(){
        unsigned long long sum = 1;
        for (int i = 2; i < 65; ++i){
            sum += square(i);
        }
        return sum;
    }
}  // namespace grains

File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n 
  | n < 1 || n> 64 = Nothing
  | otherwise = Just (go 1 1 )
    where 
      go :: Integer -> Integer -> Integer 
      go x acc 
        | x == n = acc
        | x < n = go (x+1) (2*acc)
        | x > n = error ""

total :: Integer
total = totalSum 
  where 
    totalSum = sum . map (\(Just x) ->x ). map (square) $ [1..64]